### PR TITLE
DuckDB PySpark Types

### PR DESCRIPTION
**Pull request by Tishj**
This PR extends the DataFrame creation functionality.

Now you can create DataFrames from lists of tuples.
```py
structureData = [
	(("James","","Smith"),"36636","M",3100),
	(("Michael","Rose",""),"40288","M",4300),
	(("Robert","","Williams"),"42114","M",1400),
	(("Maria","Anne","Jones"),"39192","F",5500),
	(("Jen","Mary","Brown"),"","F",-1)
]
structureSchema = StructType([
	StructField('name', StructType([
		StructField('firstname', StringType(), True),
		StructField('middlename', StringType(), True),
		StructField('lastname', StringType(), True)
		])),
	StructField('id', StringType(), True),
	StructField('gender', StringType(), True),
	StructField('salary', IntegerType(), True)
])
df2 = spark.createDataFrame(data=structureData,schema=structureSchema)
df2.show()
#┌───────────────────────────────────┬──────────────────────────────────────────────────────────────────────────────────┐
#│               name                │                                    OtherInfo                                     │
#│ struct(firstname varchar, middl…  │ struct(identifier varchar, gender varchar, salary integer, salary_grade varchar) │
#├───────────────────────────────────┼──────────────────────────────────────────────────────────────────────────────────┤
#│ {'firstname': James, 'middlenam…  │ {'identifier': 36636, 'gender': M, 'salary': 3100, 'Salary_Grade': Medium}       │
#│ {'firstname': Michael, 'middlen…  │ {'identifier': 40288, 'gender': M, 'salary': 4300, 'Salary_Grade': High}         │
#│ {'firstname': Robert, 'middlena…  │ {'identifier': 42114, 'gender': M, 'salary': 1400, 'Salary_Grade': Low}          │
#│ {'firstname': Maria, 'middlenam…  │ {'identifier': 39192, 'gender': F, 'salary': 5500, 'Salary_Grade': High}         │
#│ {'firstname': Jen, 'middlename'…  │ {'identifier': , 'gender': F, 'salary': -1, 'Salary_Grade': Low}                 │
#└───────────────────────────────────┴──────────────────────────────────────────────────────────────────────────────────┘
```

Additionally this PR also implements the PySpark type system, which stores the `DuckDBPyType` equivalent internally.
Since DuckDBs type system is more extensive than that of PySpark we have also had to add some extra type classes to the module:

These additions are:
- `Unsigned` versions of all IntegralTypes
- HugeIntegerType
- BitStringType
- TimestampNanosecondNTZType
- TimestampMilisecondNTZType
- TimestampSecondNTZType
- UUIDType
- TimeType
- TimeNTZType

Some types are still missing, namely:
- UnionType (since this is similar to StructType and StructType is somewhat complex I decided to leave it out for now)
- EnumType (similar reasoning)

### CODE REVIEW

**Review by Mytherin — APPROVED** [review]
Thanks for the PR! LGTM. Perhaps @Mause can have a look as well?

File: tools/pythonpkg/pyduckdb/spark/tests/basic/test_dataframe_read.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import pytest
+
+class TestSparkSession(object):
+	def test_read_csv_basic(self, spark):
+		pass
```
Comment (Mause) [inline tools/pythonpkg/pyduckdb/spark/tests/basic/test_dataframe_read.py:5]: Is this intentionally left blank?

File: tools/pythonpkg/pyduckdb/spark/sql/types.py [hunk 2]
```diff
@@ -1,9 +1,1221 @@
...
+
+    @classmethod
+    def typeName(cls) -> str:
+        return "timestamp_s"
+
+    def toInternal(self, dt: datetime.datetime) -> int:
+        raise ContributionsAcceptedError
+
+    def fromInternal(self, ts: int) -> datetime.datetime:
+	    raise ContributionsAcceptedError
```
Comment (Mause) [inline tools/pythonpkg/pyduckdb/spark/sql/types.py:274]: Run `black` on this file please? Indentation is important in Python

File: tools/pythonpkg/pyduckdb/spark/sql/types.py [hunk 3]
```diff
@@ -1,9 +1,1221 @@
...
+    ByteType,
+    ShortType,
+    IntegerType,
+    LongType,
+    DateType,
+    TimestampType,
+    TimestampNTZType,
+    NullType,
+]
+_all_atomic_types: Dict[str, Type[DataType]] = dict((t.typeName(), t) for t in _atomic_types)
```
Comment (Mause) [inline tools/pythonpkg/pyduckdb/spark/sql/types.py:1019]: nitpick: use dictionary comprehensions

File: tools/pythonpkg/pyduckdb/spark/sql/types.py [hunk 4]
```diff
@@ -1,9 +1,1221 @@
...
+        if len(args) > len(self):
+            raise ValueError(
+                "Can not create Row with fields %s, expected %d values "
+                "but got %s" % (self, len(self), args)
+            )
+        return _create_row(self, args)
+
+    def __getitem__(self, item: Any) -> Any:
+        if isinstance(item, (int, slice)):
+            return super(Row, self).__getitem__(item)
```
Comment (Mause) [inline tools/pythonpkg/pyduckdb/spark/sql/types.py:1176]: nitpick: This is Python 3, don't need arguments to `super()`

**Review by Mause — CHANGES_REQUESTED** [review]
Might be worth investigating using the `dataclasses` standard library module for the type classes?

Also, please move the tests into the usual test folder

File: tools/pythonpkg/pyduckdb/spark/tests/basic/test_dataframe_read.py (hunk 1 above)
Comment (Tishj) [reply tools/pythonpkg/pyduckdb/spark/tests/basic/test_dataframe_read.py:5]: It's not implemented as part of this PR yea, the whole reading/writing is big enough to warrant a separate PR I think, I'll remove the "tests" for it

File: tools/pythonpkg/pyduckdb/spark/sql/types.py (hunk 3 above)
Comment (Tishj) [reply tools/pythonpkg/pyduckdb/spark/sql/types.py:1019]: I appreciate the feedback but this is taken directly from their implementation, I'd rather not touch code that works

File: tools/pythonpkg/pyduckdb/spark/sql/dataframe.py [hunk 5]
```diff
@@ -1,29 +1,79 @@
...
+        raise NotImplementedError
+
+    def createGlobalTempView(self, name: str) -> None:
+        raise NotImplementedError
+
+    @property
+    def schema(self) -> StructType:
+        """Returns the schema of this :class:`DataFrame` as a :class:`pyspark.sql.types.StructType`.
+
+        .. versionadded:: 1.3.0
```
[the remark this answers is not in the document]
Comment (Tishj) [reply tools/pythonpkg/pyduckdb/spark/sql/dataframe.py:35]: You mean the comment? I'll update that, otherwise I don't know what you mean

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**